sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/http/policy/BearerTokenAuthenticationChallengePolicy.java

```diff
@@ -0,0 +1,164 @@
+     * @param response The Http Response containing the authentication challenge header.
+     * @return A {@link Mono} containing the status, whether the challenge was successfully extracted and handled.
+     *  if true then a follow up request needs to be sent authorized with the challenge based bearer token.
+     */
+    public Mono<Boolean> onChallengeAsync(HttpPipelineCallContext context, HttpResponse response) {
+        String authHeader = response.getHeaderValue(WWW_AUTHENTICATE);
+        if (response.getStatusCode() == 401 && authHeader != null) {
+            List<AuthenticationChallenge> challenges = parseChallenges(authHeader);
+            for (AuthenticationChallenge authenticationChallenge : challenges) {
+                Map<String, String> extractedChallengeParams =
+                    parseChallengeParams(authenticationChallenge.getChallengeParameters());
+                if (extractedChallengeParams.containsKey(CLAIMS_PARAMETER)) {
+                    String claims = new String(Base64.getUrlDecoder()
+                                                   .decode(extractedChallengeParams.get(CLAIMS_PARAMETER)));
+                    return authenticateRequest(context,
+                        () -> credential.getToken(new TokenRequestContext()
+                                                      .addScopes(scopes).setClaims(claims)), true)
+                               .flatMap(b -> Mono.just(true));
+                }
+            }
+        }
+        return Mono.just(false);
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            return Mono.error(new RuntimeException("token credentials require a URL using the HTTPS protocol scheme"));
+        }
+        HttpPipelineNextPolicy nextPolicy = next.clone();
+
+        return onBeforeRequest(context)
+                   .then(next.process())
+                   .flatMap(httpResponse -> {
+                       String authHeader = httpResponse.getHeaderValue(WWW_AUTHENTICATE);
+                       if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+                           return onChallengeAsync(context, httpResponse).flatMap(retry -> {
+                               if (retry) {
+                                   return nextPolicy.process();
+                               } else {
+                                   return Mono.just(httpResponse);
+                               }
+                           });
+                       }
+                       return Mono.just(httpResponse);
+                   });
+    }
+
+    /**
+     * Get the {@link AccessTokenCache} holding the cached access tokens and the logic to retrieve and refresh
+     * access tokens.
+     *
+     * @return the {@link AccessTokenCache}
+     */
+    public AccessTokenCache getTokenCache() {
```

Comment:
Why do we need this method? The pipeline execution will not call getTokenCache(). It only calls the `process()` method.

Reply:
Design around this method, might change in next preview.
This method is again for downstream classes, that will extend this policy, so they have access to token cache to authorize their requests.